### DISCUSSION TITLE
Compatible with Soundpipe library and is this ESP32 code about as "close to the metal" as synth programming gets?

### DISCUSSION THREAD

**Discussion by 01GOD** [marcel-licence/esp32_basic_synth#108] [General]
Thanks again for posting these files! They are great for learning synth programming! This ESP32 code is compatible with modules from the Soundpipe library and is this ESP32 code about as "close to the metal" as synth programming gets? 

Had been trying various ways to play the voltage through a speaker generated from a for loop oscillator straight from a .ino without using a DAC. Then after a lot of chatgpt assisted attempts it seemed that adding a DAC was maybe necessary. There was some way to set some pins to variable frequency PWM, but I want to know that standard C libraries would be compatible, such as Soundpipe. 

It seems that type of "for loop to oscillator" programming is known as PCM. I haven't done a lot of DSP programming directly yet though, so thanks in advance for advising if there is another term for that type. 

Been SUPER busy this week, so haven't plugged in the DAC with the ESP32 after it arrived, although hoping it will be relatively straight forward to write a for loop and add a filter algorithm from Soundpipe with a potentiometer to control the CO filter.

**Comment by marcel-licence** [comment]
You are welcome. There is actually no link between this library and Soundpipe. I can confirm that I tried to keep it as "close to the metal" as possible. There are module where you can put in samples and you get out the processes samples. The synth modules do produce samples which could be pushed into other functions.

It is possible to create a "acceptable" sound without a DAC using mechanisms like PWM. I did that with the Raspberry Pi Pico. It is also possible to use the internal DAC of the ESP32 as-well.

How do you try to link this library with Soundpipe?

Regarding PCM -> Pulse Code Modulation. As far as I understood it is just the method to put a signal into discrete value/time signal. Which is the basis of digital audio processing.

The projects, I've been made in the meantime are running in a while(true) loop. With each round a block of samples will be processed and pushed to the output (for example using I2S to DAC).

The use of Soundpipe should work to process the audio samples after generated by a synthesizer module.

**Reply by 01GOD** [reply]
I appreciate the nice answer. I think part of why I have been wanting to get to the "for loop oscillator" base level of speaker voltage control is because that way I can do high precision wave sculpture. For example, a diode clipper circuit supposedly simply clips waves. So writing that in code seems relatively easy. 

Precision wave shaping in general is interesting. Already generally tired of sloppy things. Bought some components to build a diode ladder filter, but it's not as interesting as it used to be. I like precision. 

Recommend trying some Soundpipe and Csound filter emulations. If can explain how to use those with this example, I would certainly appreciate that! For example, a filter emulation from one of those. 

Compatibility with common C libraries makes things a lot more fun pretty easily. I think this code is likely easy to use with those. I have been sprinting on some product designs while visiting a country in Asia, so I haven't tried it yet. Yesterday worked for HOURS to simply order a PCBA from a factory. 

Will be nice to have such a base level MCU synth base to sculpt precision waveforms with. Thanks again for posting the nice synth projects!

**Reply by marcel-licence** [reply]
The diode clipper example is a very interesting one. Using discrete time it will generate a different result than using an analogue circuit. In some projects I tried to avoid aliasing effects. That can also happen with the diode clipper design. In case you a looking at a single input and output value it is possible to get the same behaviour.

When I find some free time I could try to try out some stuff of Soundpipe.

I will post more in future =)

**Reply by 01GOD** [reply]
Ah, didn't know it would be much different than analog diode clipper. Have been studying diode ladder filter some. It's cool that they are little voltage controllable potentiometers basically. 

Yes, definitely notify of Soundpipe results and similar interesting things. Thanks in advance!

**Comment by 01GOD** [comment]
Past day been working on ESP32 DSP and built a simple diode clipper effect today. However, that diode ladder emulation that was supposedly authored by Will Pirkle has still seemed somewhat burly to add to a simple ESP32-S3 i2s to a DAC. Mainly because I don't know what to input to that filter emulation (what length of array, etcetera) and exactly where and then where it outputs the processed data to. Also, there are multiple editions of it for various other libraries seemingly. 

Had a look at porting that filter to ESP32 i2s yet? Oh, and if reading this, there is any filter in the repos here that sounds very similar to JP8080? 

PS: Here is the diode clipper I had used in an iPad app before. It sounded generally accurate when applied to a saw wave run through the diode ladder filter from same library.
